hilti/toolchain/src/compiler/cfg.cc

```diff
@@ -0,0 +1,893 @@
+        addEdge(last, _end);
+}
+
+Node* CFG::addGlobals(Node* parent, const Node& root) {
+    auto* p = root.parent();
+    if ( ! p )
+        return parent;
+
+    auto* mod = p->tryAs<declaration::Module>();
+    if ( ! mod )
+        return parent;
+
+    // A global variables with an init statement since they are effectively statements.
+    for ( auto* decl : mod->declarations() ) {
+        auto* global = decl->tryAs<declaration::GlobalVariable>();
+        if ( ! global )
+            continue;
+
+        if ( ! global->init() )
+            continue;
+
+        auto* stmt = getOrAddNode(global);
+        addEdge(parent, stmt);
+        parent = stmt;
+    }
+
+    return parent;
+}
+
+Node* CFG::addBlock(Node* parent, const Nodes& stmts, const Node* scope) {
+    // If `children` directly has any statements which change control flow like
+    // `throw` or `return` any statements after that are unreachable. To model
+    // such ASTs we add a flow with all statements up to the "last" semantic
+    // statement (either the last child or the control flow statement) to the
+    // CFG under `parent`. Statements after that are added as children without
+    // parents, and mixed with the previous flow.
+
+    // After this block `last` is the last reachable statement, either end of
+    // children or a control flow statement.
+    auto last = std::find_if(stmts.begin(), stmts.end(), [](auto&& c) {
+        return c && (c->template isA<statement::Return>() || c->template isA<statement::Throw>() ||
+                     c->template isA<statement::Continue>() || c->template isA<statement::Break>());
+    });
+    const bool has_dead_flow = last != stmts.end();
+    if ( has_dead_flow )
+        last = std::next(last);
+
+    // Node this block will eventually flow into.
+    auto* scope_end = getOrAddNode(create_meta_node<ScopeEnd>(scope));
+
+    // Add all statements which are part of the normal flow.
+    for ( auto&& c : (last != stmts.end() ? Nodes(stmts.begin(), last) : stmts) ) {
+        if ( ! c )
+            continue;
+
+        if ( auto&& while_ = c->tryAs<statement::While>() )
+            parent = addWhile(parent, *while_, scope_end);
+
+        else if ( auto&& for_ = c->tryAs<statement::For>() )
+            parent = addFor(parent, *for_);
+
+        else if ( auto&& if_ = c->tryAs<statement::If>() )
+            parent = addIf(parent, *if_);
+
+        else if ( auto&& try_catch = c->tryAs<statement::Try>() )
+            parent = addTryCatch(parent, *try_catch);
+
+        else if ( auto&& throw_ = c->tryAs<statement::Throw>() )
+            parent = addThrow(parent, *throw_, scope_end);
+
+        else if ( auto&& return_ = c->tryAs<statement::Return>() )
+            parent = addReturn(parent, return_->expression());
+
+        else if ( c->isA<statement::Continue>() || c->isA<statement::Break>() )
+            // `continue`/`break` statements only add flow, but no data.
+            parent = addReturn(parent, nullptr);
```

Comment:
I'm a little confused by this, since `expression` is `nullptr` this always returns `parent`, so this is just `parent = parent;`, right? Are there plans for more here?